connect-inject/handler.go

```diff
@@ -301,14 +344,53 @@ func (h *Handler) Mutate(req *v1beta1.AdmissionRequest) *v1beta1.AdmissionRespon
 		resp.PatchType = &patchType
 	}
 
+	// Check and potentially create Consul resources. This is done after
+	// all patches are created to guarantee no errors were encountered in
+	// that process before modifying the Consul cluster.
+	if h.EnableNamespaces {
+		// Check if the namespace exists. If not, create it.
+		if err := h.checkAndCreateNamespace(h.consulNamespace(pod.Namespace)); err != nil {
+			h.Log.Error("Error checking or creating namespace", "err", err,
+				"Namespace", h.consulNamespace(pod.Namespace), "Pod Name", pod.Name)
+			return &v1beta1.AdmissionResponse{
+				Result: &metav1.Status{
+					Message: fmt.Sprintf("Error checking or creating namespace: %s", err),
+				},
+			}
+		}
+
+		// For mirroring, make sure a binding rule exists if ACLs are enabled
+		if h.EnableNSMirroring && h.AuthMethod != "" {
+			if err := h.checkAndCreateBindingRule(h.consulNamespace(pod.Namespace)); err != nil {
+				h.Log.Error("Error checking or creating binding rule", "err", err,
+					"Namespace", h.consulNamespace(pod.Namespace), "Pod Name", pod.Name)
+				return &v1beta1.AdmissionResponse{
+					Result: &metav1.Status{
+						Message: fmt.Sprintf("Error checking or creating binding rule: %s", err),
+					},
+				}
+			}
+		}
+	}
+
 	return resp
 }
 
 func (h *Handler) shouldInject(pod *corev1.Pod, namespace string) (bool, error) {
-
 	// Don't inject in the Kubernetes system namespaces
-	for _, ns := range kubeSystemNamespaces {
-		if namespace == ns {
+	if kubeSystemNamespaces.Contains(namespace) {
+		return false, nil
+	}
+
+	// Namespace logic
+	if h.EnableNamespaces {
+		// If in deny list, don't inject
+		if h.DenyK8sNamespacesSet.Contains(namespace) {
```

Comment:
Short circuiting to a `true` return value for any of the checks in this method would mean that we're not checking any of the other requirements for injection. Currently, if a namespace is not in the deny list, it still might not be in the allow list and we shouldn't inject. Or it might specifically is annotated to not be injected. The deny list info is not enough to make that determination on its own.